### PR TITLE
crimson/os/seastore/omap_manager: minor bugs fixes

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/omap_manager/btree/btree_omap_manager.cc [hunk 1]
```diff
@@ -198,13 +198,17 @@ BtreeOMapManager::omap_rm_key_range(
...
-  return omap_list(
-    omap_root,
-    t,
-    first,
-    last,
-    config
-  ).si_then([this, &omap_root, &t](auto results) {
+  return seastar::do_with(
+    std::make_optional<std::string>(first),
+    std::make_optional<std::string>(last),
```
Comment (cyx1231st) [inline src/crimson/os/seastore/omap_manager/btree/btree_omap_manager.cc:203]: I think we don't need wrap first and last with do_with here, referring to the interface comments: https://github.com/ceph/ceph/blob/b66b8fde2e1cea49a8d0e0c1109092f0c378952c/src/crimson/os/seastore/omap_manager.h#L178-L179

Also, the caller has already extended the lifecycle of first and last correctly: https://github.com/ceph/ceph/blob/b66b8fde2e1cea49a8d0e0c1109092f0c378952c/src/crimson/os/seastore/seastore.cc#L1491-L1511

Comment (xxhdx1985126) [reply src/crimson/os/seastore/omap_manager/btree/btree_omap_manager.cc:203]: > I think we don't need wrap first and last with do_with here, referring to the interface comments:
> 
> https://github.com/ceph/ceph/blob/b66b8fde2e1cea49a8d0e0c1109092f0c378952c/src/crimson/os/seastore/omap_manager.h#L178-L179
> 
> Also, the caller has already extended the lifecycle of first and last correctly:
> 
> https://github.com/ceph/ceph/blob/b66b8fde2e1cea49a8d0e0c1109092f0c378952c/src/crimson/os/seastore/seastore.cc#L1491-L1511

@cyx1231st Not really, `BtreeOMapManager::omap_list()` is accepting `const std::optional<std::string> &`, so actually if we don't wrap the invocation to `BtreeOMapManager::omap_list()` here, the invocation would create a new `std::optional<std::string>` object that is alive only before the invocation is finished. However, continuations created by `BtreeOMapManager::omap_list()` will outlive those invocation, which led to a daggling reference issue.

Comment (cyx1231st) [reply src/crimson/os/seastore/omap_manager/btree/btree_omap_manager.cc:203]: Ah, I'm missing the fact that the types are actually different.

Probably we can drop the interface comments `, must alive during the call`, and also drop the unnecessary `do_with` for `first` and `last` from `SeaStore::_omap_rmkeyrange()` in this commit?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/omap_manager/btree/btree_omap_manager.cc:203]: It seems that the following code in `SeaStore::_omap_rmkeyrange()` is also capturing `first` and `last` by reference, so maybe we still need those comments.

https://github.com/ceph/ceph/blob/b66b8fde2e1cea49a8d0e0c1109092f0c378952c/src/crimson/os/seastore/seastore.cc#L1511-L1516

**Comment by xxhdx1985126** [comment]
Seems that ceph-ci's crimson build for this pr always fails: https://shaman.ceph.com/builds/ceph/wip-omap-fixes-test/120a18756663f7ed07b76e688cedd1911178f4d2/crimson/329676/

Does anyone know why? Thanks.

**Comment by xxhdx1985126** [comment]
> The following tests FAILED:
	 16 - run-tox-mgr-dashboard-openapi (Failed)
	 36 - smoke.sh (Failed)
make check failures doesn't seem related to this pr

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
jenkins test api

**Comment by cyx1231st** [comment]
> Seems that ceph-ci's crimson build for this pr always fails: https://shaman.ceph.com/builds/ceph/wip-omap-fixes-test/120a18756663f7ed07b76e688cedd1911178f4d2/crimson/329676/

SeaStore is not included in the test suite yet. It should be fine merging it without going through CI.

> Does anyone know why? Thanks.

Cross-checking the others, looks the build system is broken atm.

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
jenkins test make check

**Review by cyx1231st — APPROVED** [review]

**Pull request merged**